src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -0,0 +1,482 @@
+        Write -> backendWriteWrite key (domainTableName d) bh version txid v db
+
+backendWriteInsert :: ToJSON v => Utf8 -> Utf8 -> BlockHeight -> ReorgVersion -> TxId -> v -> Database -> IO ()
+backendWriteInsert key tn bh version txid v db =
+  exec' db ("INSERT INTO [" <> tn <> "] ('rowkey','blockheight','version','txid','tabledata') VALUES (?,?,?,?,?);")
+    [SText key
+    , SInt (fromIntegral bh)
+    , SInt (fromIntegral version)
+    , SInt (fromIntegral txid)
+    , SBlob (toStrict (Data.Aeson.encode v))]
+
+backendWriteUpdate :: ToJSON v => Utf8 -> Utf8 -> BlockHeight -> ReorgVersion -> TxId -> v -> Database -> IO ()
+backendWriteUpdate key tn bh version txid v db =
+  exec' db ("UPDATE [" <> tn <> "] SET rowkey = ?, blockheight = ?, version =  ?, txid = ?, tabledata = ? ;")
+    [SText key
+    , SInt (fromIntegral bh)
+    , SInt (fromIntegral version)
+    , SInt (fromIntegral txid)
+    , SBlob (toStrict (Data.Aeson.encode v))]
+
+-- don't need the initial select, it will just fail. Look at pact
+backendWriteWrite :: ToJSON v => Utf8 -> Utf8 -> BlockHeight -> ReorgVersion -> TxId -> v -> Database -> IO ()
+backendWriteWrite key tn bh version txid v db =
+  exec' db ("INSERT OR REPLACE INTO [" <> tn <> "] ('rowkey','blockheight','version','txid','tabledata') VALUES (?,?,?,?,?) ;")
+    [SText key, SInt (fromIntegral bh), SInt (fromIntegral version), SInt (fromIntegral txid), SBlob (toStrict (Data.Aeson.encode v))]
+
+writeUser :: WriteType -> Domain RowKey (ObjectMap PactValue) -> RowKey -> ObjectMap PactValue -> BlockHandler SQLiteEnv ()
+writeUser wt d k row = do
+      case d of
+        (UserTables _) -> do
+          BlockVersion bh version <- gets _bsBlockVersion
+          txid <- gets _bsTxId
+          userWrite k bh version txid
+  where
+    toTableName (Utf8 str) = TableName $ toS str
+    userWrite key bh version txid = do
+      olds <- readRow d key
+      case (olds, wt) of
+        (Nothing, Insert) -> ins
+        (Just _, Insert) -> internalError $ "writeUser: Insert: row not found for key " <> asString key
+        (Nothing, Write) -> ins
+        (Just old, Write) -> upd old
+        (Just old, Update) -> upd old
+        (Nothing, Update) -> internalError $ "writeUser: Update: no row found for key " <> asString key
+      where
+        upd oldrow = do
+          let row' = ObjectMap (M.union (_objectMap row) (_objectMap oldrow))
+              tn = domainTableName d
+          callDb (backendWriteUpdate (Utf8 $ toS $ asString $ key) tn bh version txid row')
+          record (prependToTableName "USER:" (toTableName $ tn)) d k row'
+        ins = do
+          let tn = domainTableName d
+          callDb $ backendWriteInsert (Utf8 $ toS $ asString $ key) tn bh version txid row
+          record (prependToTableName "USER:" (toTableName $ domainTableName d)) d k row
+
+prependToTableName :: Text -> TableName -> TableName
+prependToTableName txt (TableName name) =
+  TableName (txt <> name)
```

Comment:
Write `TableName` instances for `IsString` and `Monoid` instead?

Reply:
these are from `pact` and it already has `IsString`, not `Monoid` unfort.

Reply:
I had considered that, but then I would have to make changes to Pact. I would need the permission of @slpopejoy to do that.